conda/content_trust/__init__.py

```diff
@@ -0,0 +1,238 @@
+        path = join(context.av_data_dir, fname)
+
+        try:
+            untrusted = self._fetch_channel_signing_data(
+                context.signing_metadata_url_base,
+                KEY_MGR_FILE,
+            )
+
+            verify_delegation("key_mgr", untrusted, self.trusted_root)
+        except (ConnectionError, HTTPError) as err:
+            log.warn(err)
+        except Exception as err:
+            # TODO: more error handling
+            raise
+            log.error(err)
+        else:
+            # New key manager metadata checks out
+            trusted = untrusted
+            write_metadata_to_file(trusted, path)
+
+        # If key_mgr is unavailable from server, fall back to copy on disk
+        if not trusted and exists(path):
+            trusted = load_metadata_from_file(path)
+
+        return trusted
+
+    @staticmethod
+    def _fetch_channel_signing_data(signing_data_url, filename, etag=None, mod_stamp=None):
+        if not context.ssl_verify:
+            warnings.simplefilter("ignore", InsecureRequestWarning)
+
+        session = CondaSession()
+
+        headers = {
+            "Accept-Encoding": "gzip, deflate, compress, identity",
+            "Content-Type": "application/json",
+        }
+        if etag:
+            headers["If-None-Match"] = etag
+        if mod_stamp:
+            headers["If-Modified-Since"] = mod_stamp
+
+        try:
+            # The `auth` argument below looks a bit weird, but passing `None` seems
+            # insufficient for suppressing modifying the URL to add an Anaconda
+            # server token; for whatever reason, we must pass an actual callable in
+            # order to suppress the HTTP auth behavior configured in the session.
+            #
+            # TODO: Figure how to handle authn for obtaining trust metadata,
+            # independently of the authn used to access package repositories.
+            resp = session.get(
+                join_url(signing_data_url, filename),
+                headers=headers,
+                proxies=session.proxies,
+                auth=lambda r: r,
+                timeout=(context.remote_connect_timeout_secs, context.remote_read_timeout_secs),
+            )
+
+            resp.raise_for_status()
+        except:
+            # TODO: more sensible error handling
+            raise
+
+        # In certain cases (e.g., using `-c` access anaconda.org channels), the
+        # `CondaSession.get()` retry logic combined with the remote server's
+        # behavior can result in non-JSON content being returned.  Parse returned
+        # content here (rather than directly in the return statement) so callers of
+        # this function only have to worry about a ValueError being raised.
+        try:
+            str_data = json.loads(resp.content)
+        except json.decoder.JSONDecodeError as err:  # noqa
+            raise ValueError(f"Invalid JSON returned from {signing_data_url}/{filename}") from err
```

Comment:
Doesn't requests provide a JSON loading mechanism with advanced error handling?

Reply:
You mean `resp.json()`? That's probably a good idea, can still catch the resulting `JSONDecodeError` and use a conda exception.

Reply:
Yeah, that!